extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java

```diff
@@ -0,0 +1,226 @@
+import org.apache.druid.java.util.common.parsers.ParseException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class OpenTelemetryMetricsProtobufReader implements InputEntityReader
+{
+  private static final String VALUE_COLUMN = "value";
+
+  private final DimensionsSpec dimensionsSpec;
+  private final ByteEntity source;
+  private final String metricDimension;
+  private final String metricAttributePrefix;
+  private final String resourceAttributePrefix;
+
+  public OpenTelemetryMetricsProtobufReader(
+      DimensionsSpec dimensionsSpec,
+      ByteEntity source,
+      String metricDimension,
+      String metricLabelPrefix,
+      String resourceLabelPrefix
+  )
+  {
+    this.dimensionsSpec = dimensionsSpec;
+    this.source = source;
+    this.metricDimension = metricDimension;
+    this.metricAttributePrefix = metricLabelPrefix;
+    this.resourceAttributePrefix = resourceLabelPrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read()
+  {
+    return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+  }
+
+  List<InputRow> readAsList()
+  {
+    try {
+      return parseRequest(MetricsData.parseFrom(source.getBuffer()));
+    }
+    catch (InvalidProtocolBufferException e) {
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+  }
+
+  private List<InputRow> parseRequest(final MetricsData request)
+  {
+
+    List<ResourceMetrics> resourceMetricsList = request.getResourceMetricsList();
+    return resourceMetricsList.stream()
+            .flatMap(
+            resourceMetrics -> {
+              List<KeyValue> resourceAttributes = resourceMetrics.getResource().getAttributesList();
+              return resourceMetrics.getInstrumentationLibraryMetricsList()
+                      .stream()
+                      .flatMap(libraryMetrics -> libraryMetrics.getMetricsList()
+                              .stream()
+                              .map(metric -> parseMetric(metric, resourceAttributes))
+                              .flatMap(List::stream)
+                      );
+            }
+    ).collect(Collectors.toList());
+  }
+
+  private List<InputRow> parseMetric(Metric metric, List<KeyValue> resourceAttributes)
+  {
+
+    List<String> schemaDimensions = dimensionsSpec.getDimensionNames();
+    if (!schemaDimensions.isEmpty()) {
+      return createRows(metric, resourceAttributes, schemaDimensions::contains);
+    } else {
+      return createRows(metric, resourceAttributes, att -> !dimensionsSpec.getDimensionExclusions().contains(att));
+    }
+  }
+
+  private List<InputRow> createRows(Metric metric,
+                                    List<KeyValue> resourceAttributes,
+                                    Function<String, Boolean> attributeMembershipFn)
+  {
+
```

Comment:
Agreed, I moved that conditional up to the constructor and made the predicate a class scoped variable. 
I've inverted the call path so that the method `getDataPoints` (now called `parseMetric`) calls `parseMetric` (now calls `parseNumberDataPoint`). That seems to make more sense because in the future we'll probably need to write separate methods `parseHistogramDataPoint` and `parseSummaryDataPoint` that would use the same predicate.

Reply:
I believe there were some performance reasons for why we did it this way before. Can we check to see the performance impact?